sdk/python/lib/pulumi/automation/_cmd.py

```diff
@@ -35,46 +42,259 @@ def __str__(self) -> str:
+        self.skipVersionCheck = skipVersionCheck
+
+
+class PulumiCommand:
+    """
+    PulumiCommand manages the Pulumi CLI. It can be used to to install the CLI and run commands.
+    """
+
+    command: str
+    version: Optional[VersionInfo]
+
+    def __init__(
+        self,
+        root: Optional[str] = None,
+        version: Optional[VersionInfo] = None,
+        skip_version_check: bool = False,
+    ):
+        """
+        Creates a new PulumiCommand.
+
+        :param root: The directory where to look for the Pulumi installation. Defaults to running `pulumi` from $PATH.
+        :param version: The minimum version of the Pulumi CLI to use and validates that it is compatbile with this version.
+        :param skip_version_check: If true the version validation will be skipped.
+        """
+        self.command = os.path.join(root, "bin", "pulumi") if root else "pulumi"
+        min_version = _MINIMUM_VERSION
+        if version and version.compare(min_version) > 0:
+            min_version = version
+        current_version = (
+            subprocess.check_output([self.command, "version"]).decode("utf-8").strip()
+        )
+        if current_version.startswith("v"):
+            current_version = current_version[1:]
+        self.version = _parse_and_validate_pulumi_version(
+            min_version=min_version,
+            current_version=current_version,
+            opt_out=skip_version_check,
+        )
+
+    @classmethod
+    def install(
+        cls,
+        root: Optional[str] = None,
+        version: Optional[VersionInfo] = None,
+        skip_version_check: bool = False,
+    ) -> PulumiCommand:
+        """
+        Installs the Pulumi CLI. If root is not specified, the CLI is
+        installed to ~/.pulumi/versions/<version>. If version is not
+        specified, the version matching the SDK version is installed.
+
+        :param root: The root directory to install the CLI to.
+        :param version: The version of the CLI to install.
+        :skip_version_check: If true, the version validation will be skipped.
+        """
+        if not version:
+            version = sdk_version
+        if not root:
+            root = os.path.join(
+                os.path.expanduser("~"), ".pulumi", "versions", str(version)
+            )
+
+        try:
+            return PulumiCommand(
+                root=root, version=version, skip_version_check=skip_version_check
+            )
+        except Exception:
+            pass  # Ignore
+
+        if os.name == "nt":
+            cls._installWindows(root, version)
+        else:
+            cls._installPosix(root, version)
+
+        return PulumiCommand(
+            root=root, version=version, skip_version_check=skip_version_check
+        )
+
+    @classmethod
+    def _installWindows(cls, root: str, version: VersionInfo):
+        # TODO: Once we're on python 3.12 we can use a `with` context manager with `delete_on_close=True` here
```

Comment:
Can we do that if we need to close the filedescriptor before executing it?  Reading the docs it seems like `delete_on_close=False`, and `delete=True` would be the thing to do, then the file is deleted only when the context manager exits?

Reply:
Yes, updated the comment